packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -704,6 +705,38 @@ function shc(content: unknown): string {
     return sanitizeHtmlContentHook(content);
 }
 
+/**
+ * [ncls] - Normalize class name attribute.
+ *
+ * Transforms the provided class property value from an object/string into a string the diffing algo
+ * can operate on.
+ *
+ * This implementation is borrowed from Vue:
+ * https://github.com/vuejs/core/blob/e790e1bdd7df7be39e14780529db86e4da47a3db/packages/shared/src/normalizeProp.ts#L63-L82
+ */
+function ncls(value: unknown): string {
+    let res = '';
+
+    if (isString(value)) {
+        res = value;
+    } else if (isArray(value)) {
+        for (let i = 0; i < value.length; i++) {
+            const normalized = ncls(value[i]);
+            if (normalized) {
+                res += normalized + ' ';
+            }
+        }
```

Comment:
```suggestion
        res += ArrayMap.call(value, ncls).filter(Boolean).join(' '); 
```

Reply:
I don't have a strong opinion on this @ravijayaramappa. At this point, it is unclear to me whether or not this approach is more performant (new array allocation + extra invocations in the filter). The only thing I can say is that it is less readable than the current implementation.

@nolanlawson Any additional thoughts on this?

Reply:
Add me to the list of people without a strong opinion on this, but I will say that both versions are equally readable for me 😆

Reply:
```suggestion
        res += ArrayReduce.call(value, (acc, item) => {
            const normalized = ncls(item);
            return normalized ? `${acc} ${normalized}` : acc;
        }, '')
```
Since there are so many strong opinions, here's a third option! Should be more performant than map/filter/join, but I don't know how it compares to the for loop. It's probably the least readable, though.

Reply:
A `for` loop is always fastest. In fact we recently refactored an `Array.prototype.reduce` to a `for` loop to get some perf gains: https://github.com/salesforce/lwc/pull/3729

Normally I would not care, but this code is in the hot path, so we should prefer the `for` loop.